mesh/exodus_writer.cpp

```diff
@@ -0,0 +1,1018 @@
+
+         coordx[node_id] = coordinates[0];
+         coordy[node_id] = coordinates[1];
+
+         if (_mesh.Dimension() == 3)
+         {
+            coordz[node_id] = coordinates[2];
+         }
+      }
+   }
+   else // First-order.
+   {
+      for (int ivertex = 0; ivertex < _mesh.GetNV(); ivertex++)
+      {
+         double * coordinates = _mesh.GetVertex(ivertex);
+
+         coordx[ivertex] = coordinates[0];
+         coordy[ivertex] = coordinates[1];
+
+         if (_mesh.Dimension() == 3)
+         {
+            coordz[ivertex] = coordinates[2];
+         }
+      }
+   }
+}
+
+void ExodusIIWriter::WriteFileSize()
+{
+   // Store Exodus file size (normal==0; large==1). NB: coordinates specifed
+   // separately as components for large file.
+   const int file_size = 1;
+
+   PutAtt(NC_GLOBAL, ExodusIILabels::EXODUS_FILE_SIZE_LABEL, NC_INT, 1,
+          &file_size);
+}
+
+void ExodusIIWriter::WriteMeshDimension()
+{
+   int num_dim_id;
+   DefineDimension(ExodusIILabels::EXODUS_NUM_DIM_LABEL, _mesh.Dimension(),
+                   &num_dim_id);
+}
+
+void ExodusIIWriter::WriteNodeSets()
+{
+   // Nodesets are not currently implemented; set to zero.
+   int num_node_sets_ids;
+   DefineDimension(ExodusIILabels::EXODUS_NUM_NODE_SETS_LABEL, 0,
+                   &num_node_sets_ids);
+}
+
+void ExodusIIWriter::WriteTimesteps()
+{
+   // Set number of timesteps (ASSUME single timestep for initial verision).
+   int timesteps_dim;
+   DefineDimension(ExodusIILabels::EXODUS_TIME_STEP_LABEL, 1, &timesteps_dim);
+}
+
+void ExodusIIWriter::WriteDummyVariable()
+{
+   int dummy_var_dim_id, dummy_value = 1;
+
+   DefineDimension("dummy_var_dim", 1, &dummy_var_dim_id);
+
+   DefineAndPutVar("dummy_var", NC_INT, 1, &dummy_var_dim_id,
+                   &dummy_value);
+}
+
+void ExodusIIWriter::GenerateExodusIIElementBlocks()
+{
+   _block_ids.clear();
+   _element_ids_for_block_id.clear();
+   _element_type_for_block_id.clear();
+
+   std::unordered_set<int> observed_block_ids;
+
+   // Iterate over the elements in the mesh.
+   for (int ielement = 0; ielement < _mesh.GetNE(); ielement++)
+   {
+      Element::Type element_type = _mesh.GetElementType(ielement);
+
+      int block_id = _mesh.GetAttribute(ielement);
+
+      if (observed_block_ids.count(block_id) == 0)
+      {
+         _block_ids.push_back(block_id);
+
+         _element_type_for_block_id[block_id] = element_type;
+         _element_ids_for_block_id[block_id] = { ielement };
+
+         observed_block_ids.insert(block_id);
+      }
+      else
+      {
+         auto & block_element_ids = _element_ids_for_block_id.at(block_id);
+         block_element_ids.push_back(ielement);
+
+         // Safety check: ensure that the element type matches what we have on record
+         // for the block.
+         if (element_type != _element_type_for_block_id.at(block_id))
+         {
+            MFEM_ABORT("Multiple element types are defined for block: " << block_id);
```

Comment:
I suggest we change this so that the attempt to write an output file stops but program execution can continue. We could print a warning message stating either that mixed meshes are not yet supported or that Exodus II meshes do not support mixed element types. Whichever is more accurate.

While we're on this topic: if mixed meshes are supported but mixed blocks are not then I would like to see an outline of how we might improve such support in the future. Might it be as simple as pre-processing the mesh and dividing regions with a common attribute number into multiple blocks based on element type?

Reply:
Mixed meshes are supported. I've added two tests to check mixed first- and second-order meshes (although these are commented-out because they are dependent on  #4178 being merged (for pyramid and wedges). An ExodusII limitation is that only a single element type is supported per block. You can have multiple blocks defined each with a different element type but not a single block with multiple types. It is an error and that's why I'm using an abort. The error message could be clearer -- perhaps "Detected multiple element types on an element block. Only a single element type can be defined per block".

Yes, that would work. Provided that the ExodusII file we generate has a single element type per block we should be okay. We'd just have to be careful with propagating the new attribute numbers.

Reply:
They are marked by and integer id, yes.